File: Day #1522/Solution.java
// Concatenation of all equal-length words: sliding window per offset (0..L-1).
// Time O(|s| * L), Space O(words). Each word used exactly once.
import java.util.*;

public class Solution {
    static List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        if (words.length == 0) return res;
        int L = words[0].length(), k = words.length, n = s.length();
        int total = L * k;
        if (total > n) return res;
        Map<String, Integer> need = new HashMap<>();
        for (String w : words) need.merge(w, 1, Integer::sum);
        for (int off = 0; off < L; off++) {
            int left = off, count = 0;
            Map<String, Integer> win = new HashMap<>();
            for (int j = off; j + L <= n; j += L) {
                String w = s.substring(j, j + L);
                if (need.containsKey(w)) {
                    win.merge(w, 1, Integer::sum);
                    count++;
                    while (win.get(w) > need.get(w)) {
                        String lw = s.substring(left, left + L);
                        win.merge(lw, -1, Integer::sum);
                        left += L;
                        count--;
                    }
                    if (count == k) {
                        res.add(left);
                        String lw = s.substring(left, left + L);
                        win.merge(lw, -1, Integer::sum);
                        left += L;
                        count--;
                    }
                } else {
                    win.clear();
                    count = 0;
                    left = j + L;
                }
            }
        }
        Collections.sort(res);
        return res;
    }

    public static void main(String[] args) {
        System.out.println(findSubstring("dogcatcatcodecatdog", new String[]{"cat", "dog"}));
        System.out.println(findSubstring("barfoobazbitbyte", new String[]{"dog", "cat"}));
    }
}

File: Day #1522/solution.rs
// Concatenation of all equal-length words: sliding window per offset (0..L-1).
// Time O(|s| * L), Space O(words). Each word used exactly once.
use std::collections::HashMap;

fn find_substring(s: &str, words: &[&str]) -> Vec<usize> {
    let mut res = Vec::new();
    if words.is_empty() {
        return res;
    }
    let sb = s.as_bytes();
    let l = words[0].len();
    let k = words.len();
    let n = sb.len();
    if l * k > n {
        return res;
    }
    let mut need: HashMap<&[u8], i32> = HashMap::new();
    for w in words {
        *need.entry(w.as_bytes()).or_insert(0) += 1;
    }
    for off in 0..l {
        let mut left = off;
        let mut count = 0;
        let mut win: HashMap<&[u8], i32> = HashMap::new();
        let mut j = off;
        while j + l <= n {
            let w = &sb[j..j + l];
            if need.contains_key(w) {
                *win.entry(w).or_insert(0) += 1;
                count += 1;
                while win[w] > need[w] {
                    let lw = &sb[left..left + l];
                    *win.get_mut(lw).unwrap() -= 1;
                    left += l;
                    count -= 1;
                }
                if count == k {
                    res.push(left);
                    let lw = &sb[left..left + l];
                    *win.get_mut(lw).unwrap() -= 1;
                    left += l;
                    count -= 1;
                }
            } else {
                win.clear();
                count = 0;
                left = j + l;
            }
            j += l;
        }
    }
    res.sort();
    res
}

fn main() {
    println!("{:?}", find_substring("dogcatcatcodecatdog", &["cat", "dog"]));
    println!("{:?}", find_substring("barfoobazbitbyte", &["dog", "cat"]));
}
